lib/rubocop/ast/node_pattern/node.rb

```diff
@@ -0,0 +1,195 @@
+# frozen_string_literal: true
+
+module RuboCop
+  module AST
+    class NodePattern
+      # Base class for AST Nodes of a `NodePattern`
+      class Node < ::Parser::AST::Node
+        extend Forwardable
+        include ::RuboCop::AST::Descendence
+
+        ###
+        # To be overriden by subclasses
+        ###
+
+        def rest?
+          false
+        end
+
+        def capture?
+          false
+        end
+
+        # @return [Integer, Range] An Integer for fixed length terms, otherwise a Range.
+        # Note: `arity.end` may be `Float::INFINITY`
+        def arity
+          1
+        end
+
+        # Return [Symbol] either :ok, :raise, or :insert_wildcard
+        def in_sequence_head
+          :ok
+        end
+
+        ###
+        # Utilities
+        ###
+
+        # @return [Array<Node>]
+        def children_nodes
+          children.grep(Node)
+        end
+
+        # @return [Node] most nodes have only one child
+        def child
+          children[0]
+        end
+
+        # @return [Integer] nb of captures of that node and its descendants
+        def nb_captures
+          children_nodes.sum(&:nb_captures)
+        end
+
+        # @return [Boolean] returns true iff matches variable number of elements
+        def variadic?
+          arity.is_a?(Range)
+        end
+
+        # @return [Range] arity as a Range
+        def arity_range
+          a = arity
+          a.is_a?(Range) ? a : INT_TO_RANGE[a]
+        end
+
+        INT_TO_RANGE = Hash.new { |h, k| h[k] = k..k }
```

Comment:
Yeah, not sure it's that useful. `INT_TO_RANGE[42] # => 42..42`. I wanted to avoid returning new Ranges all the time, or having to memoize it per node, but I might do that in the end (in the constructor, coz they are frozen).